### ISSUE TITLE
Jekyll exits with 0 when a build fails

### ISSUE THREAD

**Issue by jleclanche** [jekyll/jekyll#5257]
So I created a post something like the following:

```md
---
layout: post
title: Something: Something else
---
...
```

`jekyll build` fails when parsing the title. However, it fails with **exit code 0**.

```
      Generating... 
             Error: could not read file ...: (<unknown>): mapping values are not allowed in this context at line 3 column 21
                    done in 1.752 seconds.
```

This is really bad behaviour as it means, for example, a CI build will pass despite it failing. Leaves the whole thing in a broken state.

**Comment by envygeeks** [comment]
Quote the title, you can't have a double ":" in YAML without quotes or it assumes it's parsing a k/v. Please ask non-bug user questions at https://talk.jekyllrb.com rather than on this tracker, this tracker is for bugs only.

**Comment by jleclanche** [comment]
@envygeeks That's not what the bug is about. Jekyll is **exiting with exit code 0**.

**Comment by jleclanche** [comment]
@envygeeks As I said in the original post, exiting with 0 is incorrect behaviour - can you reopen please.

**Comment by envygeeks** [comment]
Unfortunately, no.  There is no problem here as the behavior is exactly right, and expected.  You did not receive a fatal error.  Jekyll continued to process what it could and completed it's processing as per normal and exited with status 0 as you can see with the "done in n" under the error.  We only throw on fatal errors and the inability to process a file is not fatal to us so we processed what we could and gave you the error so that you could fix it.

**Comment by jleclanche** [comment]
@envygeeks this is *realy really bad* behaviour. If you want to keep that, a flag should at the very least be introduced to always exit 1 if there is any error, and not try to continue.

**Comment by envygeeks** [comment]
Remember that next time your kernel can't load a module and it continues to boot or one of your PCI devices has an error and the kernel recovers and keeps running in a sane state.  Or the next time your IDE doesn't exit when it has a non-fatal error.

**Comment by jleclanche** [comment]
@envygeeks These are the most ridiculous comparisons I've ever heard.

Don't make such behaviour default if you're concerned about backwards compatibility, but you *should* provide a flag for users who care about their sanity and the sanity of their data...

**Comment by envygeeks** [comment]
Thanks, hope you have a great day.

**Comment by Mischanix** [comment]
I agree with @jleclanche, status should definitely be non-zero when a data error occurs.  The program doesn't need to throw a fatal error when that data error occurs, it just needs to make the exit status non-zero, so that scripts which invoke jekyll can easily detect that something bad has happened.  For example, when a recoverable error occured in building, call `exit 2` at the end of the build command, where 2 is chosen because an uncaught error in ruby causes the exit status to be 1, and it's nice to be able to differentiate between levels of failure.

Exit status is a basic unix concept, and it's sad when tools, especially build tools, don't get it right.

**Comment by envygeeks** [comment]
@Mischanix A warning need not ship a bad exit code, that ***is*** a common Unix concept.  There are multiple levels of errors and our inability to read front-matter is a warning level error, a recoverable situation that does not need invoke a bad exit status.  We have done nothing wrong in our build tools and is quite common.  When apt can't read a directory that it considers inconsequential it continues on with a warning level error too, and doesn't ship a bad exit code.... we do that because at the end of the day, it did not stop us from building the hundreds of other pages... or posts.

Feel free to open another ticket to discuss the merits of such a situation (changing the state of exit codes) in a constructive manner but as it stands this ticket is staying closed because it was quickly headed to being personal and that cannot stand.

**Comment by jleclanche** [comment]
@envygeeks You closed the ticket because you didn't read it -- there was nothing personal in my original post, and there was nothing personal in any of my replies.

As for the error, it's certainly not a warning-level issue. It *fails to build the file*. Now, instead of comparing jekyll to the kernel, let me offer you a legitimate comparison: Jekyll is a build tool. Near-universally, build tools failing to build a *single* file/module/whatever will fail to build the entire project and exit immediately.

You have established behaviour which goes counter to that. Ok, fine, keep backwards compatibility if you must, but the correct behaviour should be offered behind a flag then.

I'll blame myself for not searching first, though. [This issue](https://github.com/jekyll/jekyll/issues/5104) is essentially the same bug, manifesting itself in other parts of the code. Jekyll is eating errors and not exiting appropriately because of it.

**Comment by envygeeks** [comment]
Have a lovely day @jleclanche

**Comment by benbalter** [comment]
@jleclanche 👋 stumbled across this issue when debugging a similar situation with the GitHub training team. See https://github.com/jekyll/jekyll/issues/1907#issuecomment-31548618 for some prior discussion.

@envygeeks I'm not sure that this is the intended behavior, or at least if it was purposeful, even if now intended. Doing some digging, and granted this is back in 2014, @parkr [seemed to indicate that the discussion landed on failing loudly](https://github.com/jekyll/jekyll/issues/1907#issuecomment-31587581), and then you indicated that #3211 would implement it. But, six months later [#3211 was closed without merging](https://github.com/jekyll/jekyll/pull/3211), with the opposite conclusion being given.

Looking at the blame it looks like the current behavior may be a (possible unintended) side effect of https://github.com/jekyll/jekyll/pull/4378, which was intended to target invalid dates, but seems it swallows all YAML parsing errors?

As mentioned in https://github.com/jekyll/jekyll/issues/1907#issuecomment-31548618, I'm personally torn what the best behavior is, but think, as I think you suggested, that it warrants further discussion, if there's renewed interest.

@jleclanche If you'd like to open a new issue to discuss raising errors for YAML parsing errors, I believe that's what @envygeeks was suggesting above.

/cc https://github.com/jekyll/jekyll/issues/4713

**Comment by parkr** [comment]
We made the decision here some years ago (maybe even before I joined?) to warn here. We have never failed loudly here. I shudder to think how many tickets we would receive because "all of a sudden my site didn't build OH NOES!" In certain cases, failing gracefully is the right thing to do.

For the average user, they might not look at the terminal all that much to see if any warnings happened, especially if other parts of their site are updating (i.e. the rest of the site is building). They might happen to see it when they go to shut down the `jekyll serve` process, but it may be more of a pull-your-hair-out "why isn't this page updating?!"-type experience.

I would be up for raising an error here. Unfortunately we don't have a mechanism for raising an error but continuing the process of building the rest of the site and still returning a 1 exit code. If we fail at all, we'll have to fail mid-build.
